torch/distributed/_tensor/api.py

```diff
@@ -253,6 +254,14 @@ def __tensor_unflatten__(inner_tensors, flatten_spec):
     # pyre-fixme[3]: Return type must be annotated.
     # pyre-fixme[2]: Parameter must be annotated.
     def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
+        # This is mostly for inference mode when we want to
+        # decompose CompositeImplicitAutograd ops.
+        # For the long run, we need to think of a better way to handle it.
+        if func == aten.linear.default:
```

Comment:
You actually can just remove this check for linear - `func.decompose()` will work for **all** CompositeImplicitAutograd ops, and properly return a `NotImplemented` object if it sees that the op has no `CompositeImplicitAutograd` implementation.

If you're worried about hurting eager perf - maybe it's worth a benchmark, but I think the cost of the checks that `decompose()` performs is pretty cheap.

Reply:
Yeah we should do some benchmarking, @fduwjj let's explicitly add a TODO item here, we can probably land this first and test out decompose() as follow up

Reply:
sure, let me add a TODO here and do some benchmark and then we can see if we can remove that check.